pkg/cbpf/alu_instructions.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cbpf
+
+import (
+	pb "buzzer/proto/cbpf_go_proto"
+)
+
+type Src interface {
+	pb.Reg | int32
+}
+
+func newAluInstruction[T Src](oc pb.AluOperationCode, src T) *pb.Instruction {
+	var srcType pb.SrcOperand
+	var k int32
+	switch any(src).(type) {
+	case pb.Reg:
+		srcType = pb.SrcOperand_RegSrc
+		k = int32(pb.Reg_X)
+	case int32:
+		srcType = pb.SrcOperand_Immediate
+		k = int32(src)
+	}
+	opcode := int32(0)
+
+	// The 3 least significant bits are the instruction class.
```

Comment:
Maybe add a link to the kernel documentation as a reference for the encoding format